flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,240 @@
+package net.corda.gradle.flask;
+
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import net.corda.flask.common.Flask;
+import org.gradle.api.Action;
+import org.gradle.api.NamedDomainObjectCollection;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
+import org.gradle.api.internal.file.copy.CopyAction;
+import org.gradle.api.internal.file.copy.CopyActionProcessingStream;
+import org.gradle.api.internal.file.copy.FileCopyDetailsInternal;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.BasePluginConvention;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.api.tasks.bundling.AbstractArchiveTask;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import java.io.*;
+import java.security.MessageDigest;
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class FlaskJarTask extends AbstractArchiveTask {
```

Comment:
Could you at least make this a `org.gradle.api.tasks.bundling.Zip` task? I am assuming that you are avoiding using `Jar` because of your custom manifest-writing code.

Reply:
how can I create the manifest in a Zip task? To create that manifest I need all the files that are supposed to go in the `LIB-INF` folder and that list can only be created at task execution time. I cannot find any way to add a callback that is invoked for every file that is processed in the archiving operation

Reply:
what is your concern here?

Reply:
I'm only asking you to change the base class for `FlaskJarTask` to `org.gradle.api.tasks.bundling.Zip` so that `FlaskJarTask` picks up a few of the standard "zip-related" properties that an ordinary Gradle `Jar` task would have. This should be really easy. None of your manifest-writing code should be affected, although you would need to change
```java
getArchiveExtension().convention("jar");
```
to
```java
getArchiveExtension().set("jar");
```
in order to override some configuration from the Zip super-constructor.

Reply:
```
setMetadataCharset("UTF-8");
```
is completely useless as JVMs expect JAR metadata to be always encoded using UTF-8

Reply:
I rather think that is _**precisely why**_ Gradle's own `Jar` task sets this explicitly:
```java
package org.gradle.jvm.tasks;

public class Jar extends Zip {

    public Jar() {
        this.getArchiveExtension().set("jar");
        this.setMetadataCharset("UTF-8");
        ...
    }
}
```
However, `FlaskJarTask` _doesn't_ and I think it should.

We may also conceivably want to use the `allowZip64` property.

Reply:
zip64 [is automatically handled](https://github.com/openjdk/jdk/blob/60159cff407255bfa546d98d423e8ab2e83eb0e3/src/java.base/share/classes/java/util/zip/ZipOutputStream.java#L48-L58) by the JVM since Java 7 unless explicitly prohibited (which means that the created zip files are zip64 only if that is needed by the entries and total archive size). I do not see the need to provide the users an easy way to create jar files compatible with Java 6 (especially considering that this plugin doesn't even work with Java 8)

Reply:
UTF-8 is also [the default](https://github.com/openjdk/jdk/blob/60159cff407255bfa546d98d423e8ab2e83eb0e3/src/java.base/share/classes/java/util/zip/ZipOutputStream.java#L118-L120) in the JDK (and it is not platform dependent), unless you ask for something different. Gradle, probably for legacy reason, doesn't use the standard library to create zip files but some apache library that probably has different defaults

Reply:
I'm not asking for Java 6 compatibility. I am asking that the `FlaskJarTask` be as compatible as possible (from a user's perspective) with Gradle's own `Jar` task by inheriting from `org.gradle.api.tasks.bundling.Zip`. This is trivially doable. I was under the impression that the only reason you were using `AbstractArchiveTask` in the first place was because you were originally having problems using `org.gradle.api.tasks.bundling.Jar`?

Reply:
> Gradle, probably for legacy reason, doesn't use the standard library to create zip files but some apache library that probably has different defaults.

I suspect that Gradle uses a zip library such as `commons-compress` that is more up-to-date and has more features than Java's built-in zip support, particularly relating to streaming. Even `java.util.zip.ZipFile` has more features that `ZipInputStream`.

Reply:
If I understand what you mean, then it isn't trivially doable as I have to rewrite the class completely and I don't even know how since there is the problem of getting the list of entries to generate the manifest ad task execution time, which I can't control since that would be handled by Gradle in a task of type Jar.  Otherwise if you only mean changing the parent class keeping the currently overriden `createCopyAction` method, I do not see the point of it since the additional `Jar` methods that you mentioned `setMetadataCharset` and `setZip64` would simply be ignored

Reply:
Oh, I see your problem: your `CopyAction` cannot obtain a `ZipOutputStream` from `Zip.getCompressor()` because this type is actually from a shaded internal library. OK, so we can't use that then (other than as a generic `OutputStream`) :facepalm:.

In which case, are you even supporting these properties from `AbstractArchiveTask`?
- `reproducibleFileOrder`
- `preserveFileTimestamps`
?

Reply:
for sure I'm not supporting `archivePreserveFileTimestamps` but I should be able to add support for it trivially, I think I'm supporting `reproducibleFileOrder` because I get it for free from Gradle but let me check that...

Reply:
I don't see how this task can support `reproducibleFileOrder`. Not that I think we need to, but I do think we shouldn't claim to either. Perhaps add something like this at the top of `FlaskJarTask.createCopyAction()`?
```java
if (isReproducibleFileOrder()) {
    throw new InvalidUserDataException("Not supported - very bad things etc, etc");
}
```

Reply:
I'm investingating this.. I was expecting to get it for free since the order of processing of the files is established by Gradle, according to its implementation of `FileCollection`, `FileTree` etc.. the only thing I do manually (writing the manifest, jvm argument file and Java agent file I do them in a well defined order so they should be ok). However I wrote a unit test yesterday that calls the `flaskJar` task, record the hash of the archive, then `clean` and `flaskJar` again and checks the two archives hashes and I noticed it fails randomly... (I also took care of adding `reproducibleFileOrder` and `archivePreserveFileTimestamps` to all of the included archives in `LIB-INF` built by Gradle in the same project)

I'm trying to understand which part isn't deterministic...

Reply:
also, while the sha256 was different, the output of `jar tvf flask.jar` was identical for both (I mean **really** identical, with `diff` showing no output). I'm wondering if the deflate compression algorithm contains any undeterminism

Reply:
Personally, I'd assume that it isn't deterministic unless something actively makes it so. Perhaps look to see what the existing `Zip` task does? But as I said, this isn't _really_ a problem so long as we tell the users that this setting this property will not do what they expect.

Reply:
> also, while the sha256 was different, the output of `jar tvf flask.jar` was identical for both (I mean **really** identical, with `diff` showing no output). I'm wondering if the deflate compression algorithm contains any undeterminism

Where the `zipinfo` outputs also identical? The zip format also contains the `fileMode` and `directoryMode` of its entries, and while Java's `ZipInputStream` and `ZipOutputStream` don't support these attributes, I think Gradle's internal zip library does...

Reply:
the problem is that `Properties.store` writes the current timestamp as a first line in the property file, as in 
```
#Fri Feb 12 14:46:13 CET 2021
0=843e6bec94c3e4f4277a66c8f6896ca3cfb566ad0f4d52844676a5d2ed3a0a1b\=testArgument
```
[there seems](https://github.com/openjdk/jdk/blob/05a764f4ffb8030d6b768f2d362c388e5aabd92d/src/java.base/share/classes/java/util/Properties.java#L926-L927
) to be no way to stop it from doing that 🤦🏿

Reply:
I've replaced `Properties.store` with `Properties.storeToXML` since that doesn't write the timestamp and gives us the full determinism. I'll now dig into Gradle source code to make sure this is actually true for the order of processing of files in `FileCollection`, but as of now, the unit test comparing archives generates one after the other is green. We might decide to disable it if this proves fragile (I mean, after the changes to the way properties are stored I haven't detected any other source of indeterminism, but who knows...)

Reply:
OK, but there's no need to spend too long on this because we don't need the plugin to do this yet. The `reproducibleFileOrder` property is only about the making the `ordering` of the entries repeatable anyway.

Reply:
ok, btw I confirm that my task supports `reproducibleFileOrder` since that depends entirely on Gradle's [CopyActionExecuter](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/tasks/bundling/AbstractArchiveTask.java#L501-L507) that my task still uses